integration/conftest.py

```diff
@@ -151,78 +159,74 @@ def tahoe_venv(request, reactor):
+        return ""
 
-        print("Dumping {} flogtool logfiles to '{}'".format(len(flogs), flog_file))
+    with start_task(action_type=u"integration:flog_gatherer"):
+        out_protocol = _CollectOutputProtocol()
+        gather_dir = join(temp_dir, 'flog_gather')
         reactor.spawnProcess(
-            flog_protocol,
+            out_protocol,
             flog_binary,
             (
-                'flogtool', 'dump', join(temp_dir, 'flog_gather', flogs[0])
-            ),
+                u'flogtool', u'create-gatherer',
+                u'--location', u'tcp:localhost:3117',
+                u'--port', u'3117',
+                gather_dir,
+            )
         )
-        print("Waiting for flogtool to complete")
-        try:
-            pytest_twisted.blockon(flog_protocol.done)
-        except ProcessTerminated as e:
-            print("flogtool exited unexpectedly: {}".format(str(e)))
-        print("Flogtool completed")
+        pytest_twisted.blockon(out_protocol.done)
 
-    request.addfinalizer(cleanup)
+        magic_text = "Gatherer waiting at"
+        executable = str(tahoe_venv.joinpath('bin', 'twistd'))
+        args = (
+            'twistd', '--nodaemon', '--python',
+            join(gather_dir, 'gatherer.tac'),
+        )
+        action_fields = {
+            "action_type": u"integration:flog-gatherer:service",
+        }
+        pytest_twisted.blockon(
+            run_service(reactor, request, action_fields, magic_text, executable, args, cwd=gather_dir)
+        )
 
-    with open(join(gather_dir, 'log_gatherer.furl'), 'r') as f:
-        furl = f.read().strip()
-    return furl
+        def cleanup():
+            flog_file = 'integration.flog_dump'
+            flog_protocol = _DumpOutputProtocol(open(flog_file, 'w'))
+            flogs = [x for x in listdir(gather_dir) if x.endswith('.flog')]
+
+            print("Dumping {} flogtool logfiles to '{}'".format(len(flogs), flog_file))
+            reactor.spawnProcess(
+                flog_protocol,
+                flog_binary,
+                (
+                    'flogtool', 'dump', join(gather_dir, flogs[0])
+                ),
+            )
+            print("Waiting for flogtool to complete")
+            try:
+                pytest_twisted.blockon(flog_protocol.done)
+            except ProcessTerminated as e:
+                print("flogtool exited unexpectedly: {}".format(str(e)))
+            print("Flogtool completed")
+
+        request.addfinalizer(cleanup)
+
+        with open(join(gather_dir, 'log_gatherer.furl'), 'r') as f:
+            furl = f.read().strip()
+        return furl
 
 
 @pytest.fixture(scope='session')
-@log_call(
-    action_type=u"integration:introducer",
-    include_args=["temp_dir", "flog_gatherer"],
-    include_result=False,
-)
 def introducer(reactor, tahoe_venv, temp_dir, flog_gatherer, request):
-    config = '''
+    with start_task(action_type=u"integration:introducer").context():
```

Comment:
I can't recall a time when I've used `start_task` instead of `start_action`.  I understand that the difference is that `start_task` will not search out a parent from the Eliot context: it will always be a new top-level action?  So "running the introducer" is a top-level action regardless of the Eliot context in which this fixture is invoked?

I'm not sure how to reason about when it makes sense to start a new top-level action (apart from there being no other active action) and when it makes sense to respect the current Eliot context.  Do you have a set of rules for this you can share?  Or maybe just a comment about this specific case.

Reply:
Mostly the reason it is this way is that I played around with various options, looking at the eliot-tree output of test runs with them.
- Partly what is going on is that fixtures are run outside the context of the thing requesting them, particularly session fixtures. One iteration did have a session-scope auto-use fixture that did have a task, that all other session scoped fixtures were child actions of, but that didn't make it easier to read the logs. I *do* have that for tests, which does work better (though I'm not sure there are fest (i.e. function) scoped fixtures that would capture logs for.
- If you have multiple, mostly unrelated fixtures, that each have an action wrapping them, then each action is going to be a child of the previous one, but the nesting is going to be somewhat arbitrary, and makes the logs harder to read. (One particular example of confusing nesting, in a slightly different context, is that the magic-folder (stdout) logs from restarting the service where in the test that restarted them, even if it kept running into a later test).